include/dice/template-library/standard_layout_tuple.hpp

```diff
@@ -0,0 +1,158 @@
+#ifndef DICE_TEMPLATELIBRARY_STANDARDLAYOUTTUPLE_HPP
+#define DICE_TEMPLATELIBRARY_STANDARDLAYOUTTUPLE_HPP
+
+#include <dice/template-library/type_list.hpp>
+#include <dice/template-library/type_traits.hpp>
+
+#include <cstddef>
+#include <functional>
+#include <type_traits>
+#include <utility>
+
+namespace dice::template_library {
+
+	/**
+	 * A tag type to tell a tuple to construct all fields with the same arguments.
+	 *
+	 * @example
+	 * @code
+	 * standard_layout_tuple<int, int, int> const uniform{uniform_construct, 42};
+	 * assert(uniform == standard_layout_tuple<int, int, int>{42, 42, 42});
+	 * @endcode
+	 */
+	struct uniform_construct_t {};
+	inline constexpr uniform_construct_t uniform_construct;
+
+	/**
+	 * A tag type to tell a tuple to construct each member with one of the given arguments.
+	 * This mostly exists for backwards compatibility purposes and clarity. It is the same as the regular constructor.
+	 *
+	 * @example
+	 * @code
+	 * standard_layout_tuple<int, int, int> const individual{individual_construct, 1, 2, 3};
+	 * assert(individual == standard_layout_tuple<int, int, int>{1, 2, 3});
+	 * @encode
+	 */
+	struct individual_construct_t {};
+	inline constexpr individual_construct_t individual_construct;
+
+	/**
+	 * A tuple that fulfills std::is_standard_layout
+	 */
+	template<typename ...Ts>
+	struct standard_layout_tuple;
+
+	template<>
+	struct standard_layout_tuple<> {
+		constexpr standard_layout_tuple() noexcept = default;
+		explicit constexpr standard_layout_tuple(individual_construct_t) noexcept {}
+
+		template<typename ...Args>
+		explicit constexpr standard_layout_tuple(uniform_construct_t, [[maybe_unused]] Args &&...args) noexcept {}
+
+		static constexpr size_t size() noexcept {
+			return 0;
+		}
+
+		template<size_t offset, size_t count>
+		[[nodiscard]] standard_layout_tuple const &subtuple() const noexcept {
+			static_assert(offset == 0 && count == 0);
+			return *this;
+		}
+
+		constexpr auto operator<=>(standard_layout_tuple const &other) const noexcept = default;
+		constexpr bool operator==(standard_layout_tuple const &other) const noexcept = default;
+	};
+
+	template<typename T, typename ...Ts>
+	struct standard_layout_tuple<T, Ts...> {
+	private:
+		T first;
+		[[no_unique_address]] standard_layout_tuple<Ts...> rest;
```

Comment:
why do we have to use `no_unique_address`? is it to avoid storing anything for `standard_layout_tuple<>`?

Reply:
already resolved